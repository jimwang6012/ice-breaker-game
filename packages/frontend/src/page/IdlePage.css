.music1 {
  width: 100%;

  position: absolute;
  height: 100%;
  opacity: 0;
}
