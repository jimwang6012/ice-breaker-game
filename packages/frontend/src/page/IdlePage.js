import React, { useContext, useState, useEffect, useRef } from "react";
import { useNavigate } from "react-router-dom";
import { AppContext } from "../AppContextProvider";
import socket from "../Socket";
import { MainButton } from "../component/MainButton";
import { Modal, NumberInput, Space, Group, Tooltip } from "@mantine/core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faClipboard,
  faVolumeHigh,
  faVolumeMute,
} from "@fortawesome/free-solid-svg-icons";
import { ColorSwatchGroup } from "../component/ColorSwatchGroup";
import { PlayerSelector } from "../component/PlayerSelector";

export default function IdlePage() {
  const navigate = useNavigate();
  const toHome = () => {
    socket.disconnect();
    navigate("/home");
  };
  const {
    isHost,
    setIsHost,
    setHostId,
    roomId,
    players,
    handlers,
    config,
    setConfig,
    colors,
    colorStatus,
    setColorStatus,
    playing,
    toggle,
    stop,
  } = useContext(AppContext);

  const [opened, setOpened] = useState(false);
  const [isUpdateResponse, setUpdateResponse] = useState(false);
  const [isSuccess, setSuccess] = useState(false);
  const [checkedColor, setCheckedColor] = useState(null);
  const [isReady, setIsReady] = useState(false);
  const [isReadyFail, setIsReadyFail] = useState(false);

  const refRoomSize = useRef(null);
  const refBoardSize = useRef(null);
  const refRoundTime = useRef(null);
  const refBreakTime = useRef(null);
  const refBreakerPlayer = useRef(null);

  const updateConfig = () => {
    const newConfig = {
      roomSize: refRoomSize.current.value,
      boardSize: refBoardSize.current.value,
      roundTime: refRoundTime.current.value,
      breakTime: refBreakTime.current.value,
      breakerName: refBreakerPlayer.current.value,
    };
    socket.emit("update-config", { roomId, config: newConfig }, (isSuccess) => {
      setUpdateResponse(true);
      setSuccess(isSuccess);
    });
  };

  const playerReady = () => {
    socket.emit(
      "player-ready",
      { roomId, playerId: socket.id, checkedColor },
      (isReady) => {
        if (isReady) {
          setIsReady(true);
          setIsReadyFail(false);
        } else {
          setIsReadyFail(true);
        }
      }
    );
  };

  const playerUnReady = () => {
    setIsReady(false);
    socket.emit("player-unready", { roomId, playerId: socket.id });
  };

  const startGame = () => {
    socket.emit("start-game", { roomId }, (isSuccess) => {
      if (!isSuccess) {
        alert("Not everyone is ready");
      }
    });
  };

  const checkColorBox = (index) => {
    if (index === checkedColor) {
      setCheckedColor(null);
    } else {
      setCheckedColor(index);
    }
  };

  useEffect(() => {
    playing ? toggle() : stop();
    socket.emit("room-information", { roomId }, (room) => {
      handlers.setState(room.players);
      setColorStatus(room.colorStatus);
      setHostId(room.hostId);
    });
    const startGame = (game) => {
      if (game.hostId === socket.id) {
        setIsHost(true);
      } else {
        setIsHost(false);
      }
      if (game?.config) setConfig(game.config);
      if (game?.colorStatus) setColorStatus(game.colorStatus);
      if (game.board) navigate(`/${roomId}/game`, { state: { game: game } });
    };
    socket.on("game-update", startGame);
    return () => {
      socket.off("game-update", startGame);
    };
  }, []);

  return (
    <div>
      <button
        onClick={playing ? stop : toggle}
        className="absolute z-10 px-5 py-3 font-semibold text-white rounded-lg top-2 left-2 text-l bg-ice-6 hover:bg-ice-5"
      >
        <FontAwesomeIcon icon={playing ? faVolumeHigh : faVolumeMute} />
      </button>
      <div className="flex items-center h-screen bg-ice-8 opacity-90">
        <Modal
          centered
          opened={opened}
          onClose={() => setOpened(false)}
          title="Game Config"
        >
          <NumberInput
            label="Room Size"
            description="From 0 to 9, greater than current player count"
            placeholder="Number of players"
            defaultValue={config.roomSize}
            min={0}
            max={9}
            ref={refRoomSize}
            disabled={!isHost}
            styles={{
              input: {
                "&:disabled": {
                  backgroundColor: "white",
                  color: "black",
                },
              },
            }}
          />
          <NumberInput
            label="Board Size"
            description="From 5 to 20"
            placeholder="Size of game board"
            defaultValue={config.boardSize}
            min={0}
            max={20}
            ref={refBoardSize}
            disabled={!isHost}
            styles={{
              input: {
                "&:disabled": {
                  backgroundColor: "white",
                  color: "black",
                },
              },
            }}
          />
          <NumberInput
            label="Round Time"
            description="Unit is in seconds"
            placeholder="Round time limit"
            defaultValue={config.roundTime}
            min={0}
            ref={refRoundTime}
            disabled={!isHost}
            styles={{
              input: {
                "&:disabled": {
                  backgroundColor: "white",
                  color: "black",
                },
              },
            }}
          />
          <NumberInput
            label="Break Delay"
            description="From 0.0 to 1.0, unit is in seconds"
            placeholder="Tile break delay"
            defaultValue={config.breakTime}
            min={0}
            max={1}
            precision={1}
            ref={refBreakTime}
            disabled={!isHost}
            styles={{
              input: {
                "&:disabled": {
                  backgroundColor: "white",
                  color: "black",
                },
              },
            }}
          />
          <PlayerSelector
            players={players}
            defaultValue={config.breakerName}
            breakerRef={refBreakerPlayer}
            disabled={!isHost}
          />
          {isHost && (
            <>
              <div className="flex justify-center">
                <button
                  onClick={updateConfig}
                  className="px-5 py-3 mt-5 font-semibold text-white rounded-lg text-l bg-ice-6 hover:bg-ice-5"
                >
                  Update Config
                </button>
              </div>
              <Space h="xs" />
            </>
          )}
          {isUpdateResponse ? (
            isSuccess ? (
              <p className="text-center text-green-500">
                Successfully updated game config
              </p>
            ) : (
              <p className="text-center text-red-500">
                Failed to update game config
              </p>
            )
          ) : (
            <p className="invisible text-center">Invisible text</p>
          )}
        </Modal>
        {/* Room Info */}
        <div className="flex flex-col items-center justify-center w-full gap-20 pt-5 ">
          <div className="text-6xl font-bold text-white">
            {isHost ? "Invite your friends!" : "Wait for host to Start!"}
          </div>
          {/* Room Code  */}
          <Group
            position="center"
            grow
            style={{
              borderWidth: 3,
              borderRadius: 10,
              borderColor: isReady ? "#003853" : "transparent",
            }}
          >
            <div className="w-2/5 py-16 rounded-lg shadow-lg px-28 bg-ice-2 ">
              <div className="flex flex-row justify-center gap-2 text-2xl font-bold text-black">
                Room Code: {roomId}
                <button onClick={() => navigator.clipboard.writeText(roomId)}>
                  <FontAwesomeIcon
                    className="hover:text-ice-7"
                    icon={faClipboard}
                  />
                </button>
              </div>
              <Space h="xl" />
              <Space h="xl" />
              <Space h="xs" />
              {/* Color Picker */}
              <Group
                direction="row"
                spacing="lg"
                align="center"
                position="center"
              >
                <ColorSwatchGroup
                  colorStrings={colors}
                  colorStatus={colorStatus}
                  colorChecked={checkedColor}
                  isReady={isReady}
                  handleClick={checkColorBox}
                />
                <Tooltip
                  opened={isReadyFail}
                  label={"Please choose a valid color"}
                  position="right"
                >
                  <MainButton
                    handleClick={isReady ? playerUnReady : playerReady}
                    text={isReady ? "Cancel" : "Ready"}
                  />
                </Tooltip>
              </Group>
            </div>
          </Group>
          {isHost ? (
            <div className="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2">
              {/* Two Buttons */}
              <MainButton
                handleClick={() => {
                  toHome();
                  stop();
                }}
                text="Leave Room"
              />
              <MainButton handleClick={startGame} text="Start Game" />
              <MainButton
                handleClick={() => {
                  setUpdateResponse(false);
                  setOpened(true);
                }}
                text="Game Config"
              />
            </div>
          ) : (
            <div className="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2">
              <MainButton handleClick={toHome} text="Leave Room" />
              <MainButton
                handleClick={() => {
                  setUpdateResponse(false);
                  setOpened(true);
                }}
                text="Game Config"
              />
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
