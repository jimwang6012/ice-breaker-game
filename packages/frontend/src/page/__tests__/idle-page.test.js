import { BrowserRouter, Router } from "react-router-dom";
import IdlePage from "../IdlePage";
import { render } from "@testing-library/react";

it("renders correctly", () => {
  const tree = render(
    <BrowserRouter>
      <IdlePage />
    </BrowserRouter>
  );
  expect(tree).toMatchSnapshot();
});
