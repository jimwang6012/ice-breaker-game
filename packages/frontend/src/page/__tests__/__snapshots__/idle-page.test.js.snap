// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          class="absolute z-10 px-5 py-3 font-semibold text-white rounded-lg top-2 left-2 text-l bg-ice-6 hover:bg-ice-5"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-volume-high "
            data-icon="volume-high"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M412.6 182c-10.28-8.334-25.41-6.867-33.75 3.402c-8.406 10.24-6.906 25.35 3.375 33.74C393.5 228.4 400 241.8 400 255.1c0 14.17-6.5 27.59-17.81 36.83c-10.28 8.396-11.78 23.5-3.375 33.74c4.719 5.806 11.62 8.802 18.56 8.802c5.344 0 10.75-1.779 15.19-5.399C435.1 311.5 448 284.6 448 255.1S435.1 200.4 412.6 182zM473.1 108.2c-10.22-8.334-25.34-6.898-33.78 3.34c-8.406 10.24-6.906 25.35 3.344 33.74C476.6 172.1 496 213.3 496 255.1s-19.44 82.1-53.31 110.7c-10.25 8.396-11.75 23.5-3.344 33.74c4.75 5.775 11.62 8.771 18.56 8.771c5.375 0 10.75-1.779 15.22-5.431C518.2 366.9 544 313 544 255.1S518.2 145 473.1 108.2zM534.4 33.4c-10.22-8.334-25.34-6.867-33.78 3.34c-8.406 10.24-6.906 25.35 3.344 33.74C559.9 116.3 592 183.9 592 255.1s-32.09 139.7-88.06 185.5c-10.25 8.396-11.75 23.5-3.344 33.74C505.3 481 512.2 484 519.2 484c5.375 0 10.75-1.779 15.22-5.431C601.5 423.6 640 342.5 640 255.1S601.5 88.34 534.4 33.4zM301.2 34.98c-11.5-5.181-25.01-3.076-34.43 5.29L131.8 160.1H48c-26.51 0-48 21.48-48 47.96v95.92c0 26.48 21.49 47.96 48 47.96h83.84l134.9 119.8C272.7 477 280.3 479.8 288 479.8c4.438 0 8.959-.9314 13.16-2.835C312.7 471.8 320 460.4 320 447.9V64.12C320 51.55 312.7 40.13 301.2 34.98z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="flex items-center h-screen bg-ice-8 opacity-90"
        >
          <div
            class="flex flex-col items-center justify-center w-full gap-20 pt-5 "
          >
            <div
              class="text-6xl font-bold text-white"
            >
              Wait for host to Start!
            </div>
            <div
              class="mantine-Group-root mantine-1u7f61y"
              style="border-width: 3px; border-radius: 10px; border-color: transparent;"
            >
              <div
                class="mantine-1vp280o mantine-Group-child w-2/5 py-16 rounded-lg shadow-lg px-28 bg-ice-2 "
              >
                <div
                  class="flex flex-row justify-center gap-2 text-2xl font-bold text-black"
                >
                  Room Code: 
                  
                  <button>
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-clipboard hover:text-ice-7"
                      data-icon="clipboard"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 384 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M336 64h-53.88C268.9 26.8 233.7 0 192 0S115.1 26.8 101.9 64H48C21.5 64 0 85.48 0 112v352C0 490.5 21.5 512 48 512h288c26.5 0 48-21.48 48-48v-352C384 85.48 362.5 64 336 64zM192 64c17.67 0 32 14.33 32 32c0 17.67-14.33 32-32 32S160 113.7 160 96C160 78.33 174.3 64 192 64zM272 224h-160C103.2 224 96 216.8 96 208C96 199.2 103.2 192 112 192h160C280.8 192 288 199.2 288 208S280.8 224 272 224z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  class="mantine-1eabysl"
                />
                <div
                  class="mantine-1eabysl"
                />
                <div
                  class="mantine-10dh38u"
                />
                <div
                  class="mantine-Group-root mantine-3suwep"
                >
                  <div
                    class="mantine-Group-root mantine-1u7f61y"
                  >
                    <div
                      class="mantine-SimpleGrid-root mantine-Group-child mantine-1qxm3qj"
                    />
                  </div>
                  <div
                    class="mantine-Tooltip-root mantine-Group-child mantine-1so9ne2"
                  >
                    <div
                      class="flex"
                    >
                      <button
                        class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                      >
                        Ready
                      </button>
                      <div
                        class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                      >
                        Ready
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2"
            >
              <div
                class="flex"
              >
                <button
                  class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                >
                  Leave Room
                </button>
                <div
                  class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                >
                  Leave Room
                </div>
              </div>
              <div
                class="flex"
              >
                <button
                  class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                >
                  Game Config
                </button>
                <div
                  class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                >
                  Game Config
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        dir="ltr"
        style="position: relative; z-index: 200;"
      />
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        class="absolute z-10 px-5 py-3 font-semibold text-white rounded-lg top-2 left-2 text-l bg-ice-6 hover:bg-ice-5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-volume-high "
          data-icon="volume-high"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M412.6 182c-10.28-8.334-25.41-6.867-33.75 3.402c-8.406 10.24-6.906 25.35 3.375 33.74C393.5 228.4 400 241.8 400 255.1c0 14.17-6.5 27.59-17.81 36.83c-10.28 8.396-11.78 23.5-3.375 33.74c4.719 5.806 11.62 8.802 18.56 8.802c5.344 0 10.75-1.779 15.19-5.399C435.1 311.5 448 284.6 448 255.1S435.1 200.4 412.6 182zM473.1 108.2c-10.22-8.334-25.34-6.898-33.78 3.34c-8.406 10.24-6.906 25.35 3.344 33.74C476.6 172.1 496 213.3 496 255.1s-19.44 82.1-53.31 110.7c-10.25 8.396-11.75 23.5-3.344 33.74c4.75 5.775 11.62 8.771 18.56 8.771c5.375 0 10.75-1.779 15.22-5.431C518.2 366.9 544 313 544 255.1S518.2 145 473.1 108.2zM534.4 33.4c-10.22-8.334-25.34-6.867-33.78 3.34c-8.406 10.24-6.906 25.35 3.344 33.74C559.9 116.3 592 183.9 592 255.1s-32.09 139.7-88.06 185.5c-10.25 8.396-11.75 23.5-3.344 33.74C505.3 481 512.2 484 519.2 484c5.375 0 10.75-1.779 15.22-5.431C601.5 423.6 640 342.5 640 255.1S601.5 88.34 534.4 33.4zM301.2 34.98c-11.5-5.181-25.01-3.076-34.43 5.29L131.8 160.1H48c-26.51 0-48 21.48-48 47.96v95.92c0 26.48 21.49 47.96 48 47.96h83.84l134.9 119.8C272.7 477 280.3 479.8 288 479.8c4.438 0 8.959-.9314 13.16-2.835C312.7 471.8 320 460.4 320 447.9V64.12C320 51.55 312.7 40.13 301.2 34.98z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="flex items-center h-screen bg-ice-8 opacity-90"
      >
        <div
          class="flex flex-col items-center justify-center w-full gap-20 pt-5 "
        >
          <div
            class="text-6xl font-bold text-white"
          >
            Wait for host to Start!
          </div>
          <div
            class="mantine-Group-root mantine-1u7f61y"
            style="border-width: 3px; border-radius: 10px; border-color: transparent;"
          >
            <div
              class="mantine-1vp280o mantine-Group-child w-2/5 py-16 rounded-lg shadow-lg px-28 bg-ice-2 "
            >
              <div
                class="flex flex-row justify-center gap-2 text-2xl font-bold text-black"
              >
                Room Code: 
                
                <button>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-clipboard hover:text-ice-7"
                    data-icon="clipboard"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 384 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M336 64h-53.88C268.9 26.8 233.7 0 192 0S115.1 26.8 101.9 64H48C21.5 64 0 85.48 0 112v352C0 490.5 21.5 512 48 512h288c26.5 0 48-21.48 48-48v-352C384 85.48 362.5 64 336 64zM192 64c17.67 0 32 14.33 32 32c0 17.67-14.33 32-32 32S160 113.7 160 96C160 78.33 174.3 64 192 64zM272 224h-160C103.2 224 96 216.8 96 208C96 199.2 103.2 192 112 192h160C280.8 192 288 199.2 288 208S280.8 224 272 224z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="mantine-1eabysl"
              />
              <div
                class="mantine-1eabysl"
              />
              <div
                class="mantine-10dh38u"
              />
              <div
                class="mantine-Group-root mantine-3suwep"
              >
                <div
                  class="mantine-Group-root mantine-1u7f61y"
                >
                  <div
                    class="mantine-SimpleGrid-root mantine-Group-child mantine-1qxm3qj"
                  />
                </div>
                <div
                  class="mantine-Tooltip-root mantine-Group-child mantine-1so9ne2"
                >
                  <div
                    class="flex"
                  >
                    <button
                      class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
                    >
                      Ready
                    </button>
                    <div
                      class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
                    >
                      Ready
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col items-center justify-around w-3/5 xl:flex-row gap-y-2"
          >
            <div
              class="flex"
            >
              <button
                class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
              >
                Leave Room
              </button>
              <div
                class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
              >
                Leave Room
              </div>
            </div>
            <div
              class="flex"
            >
              <button
                class="absolute z-40 px-6 py-3 text-xl font-semibold text-white rounded-lg shadow-inner bg-ice-6 hover:mt-1 active:mt-2"
              >
                Game Config
              </button>
              <div
                class="px-6 py-3 mt-2 text-xl font-semibold rounded-lg text-ice-7 bg-ice-7"
              >
                Game Config
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
